freestyle/shared/src/main/scala/freestyle/module.scala

```diff
@@ -16,171 +16,190 @@
-          val tyDefs = algebras.zipWithIndex.drop(2).map { case (alg, pos) =>
-            q"type ${ccs(pos)}[$AA] = Coproduct[$alg.Op, ${ccs(pos-1)}, $AA]"
-          }
-          val opDef = q"type Op[$AA] = ${ccs.last}[$AA]"
-          tyDef1 :: (tyDefs :+ opDef)
-      }
+  private[this] implicit final class ValidateFlatmap[E, A](v: Validated[E, A]) {
+    def flatMap[EE >: E, B](f: A => Validated[EE, B]): Validated[EE, B] =
+      v.andThen(f)
+  }
 
-    // findAlgebras starts from the ClassSymbol of the `@moudle-annotated trait
-    val algebras   = findAlgebras(weakTypeOf[A].typeSymbol.companion.asClass)
-    val coproducts = mkCoproduct(algebras)
-    //ugly hack because, as String,  it does not typecheck, early which we need for types to be in scope
-    val parsed     = coproducts.map( cop => c.parse(cop.toString))
-
-    val expr = if (algebras.length >= 2)
-      q"""new {
-        import _root_.cats.data.Coproduct
-        ..$parsed
-      }"""
+  def impl(annottees: c.Expr[Any]*): Tree =
+    foldAbort(for {
+      cls             <- decodeAnnotationTarget(annottees.toList)
+      effectTuples    <- gatherEffects(cls)
+      classTree       <- makeClassTree(cls)
+      objectTree      <- makeObjectTree(cls, effectTuples)
+    } yield q"$classTree; $objectTree")
+
+  def computeCoproductImpl: Tree =
+    foldAbort(for {
+      tpe             <- compassionateCompanionTypeOf(c.internal.enclosingOwner.owner)
+      fullEffectTypes =  expandEffectType(tpe)
+      fullEffectTrees <- fullEffectTypes.traverse(typeToTypeTree)
+      //coproductTree   <- makeCatsCoproduct(fullEffectTrees) // TODO: make this configurable?
+      coproductTree   <- makeIotaCoproduct(fullEffectTrees)
+    } yield q"""new { ..$coproductTree }""")
+
+  def compassionateCompanionTypeOf(sym: Symbol): ValidatedNel[String, Type] =
+    if (sym.companion.isType)
+      sym.companion.asType.toType.validNel
     else
-      q"""new {
-        ..$parsed
-      }"""
-    c.Expr[Any](expr)
+      s"unable to find companion for $sym when expanding coproduct".invalidNel
+
+  private[this] def decodeAnnotationTarget(annottees: List[c.Expr[Any]]) =
+    annottees.map(_.tree) match {
+      case (cls: ClassDef) :: Nil =>
+        if (cls.mods.hasFlag(Flag.TRAIT | Flag.ABSTRACT)) cls.validNel
+        else
+          "The `@module` annotation can only be applied to a trait or an abstract class.".invalidNel
+
+      case (cls: ClassDef) :: more =>
+        "The trait or class annotated with `@module` must have no companion object.".invalidNel
+      case trees =>
+        "Unexpected trees $trees encountered for `@module` annotation".invalidNel
   }
 
-}
-
-object moduleImpl {
-
-  def impl(c: Context)(annottees: c.Expr[Any]*): c.universe.Tree = {
-    import c.universe._
-    import c.universe.internal.reificationSupport._
+  private[this] def gatherEffects(cls: ClassDef): ValidatedNel[String, List[(Name, Tree)]] =
```

Comment:
Unless I'm wrong, there seems not to be any case in this procedure that would return an _Invalid_ result. That being so, we could just return the `List`, and just write into the `collect [...] => (valDef.name, valDef.tpt)`.